Include reference scripts size in min fee calculation

teodanciu opened the pull request:
# Description

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

Resolves https://github.com/IntersectMBO/cardano-ledger/issues/3952

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

teodanciu committed:
Add`originalBytesSize` to `SafeToHash` class with default implementation

teodanciu committed:
Add `minFeeRefScriptCoinsPerByte` to Conway PParams

teodanciu committed:
Add test to verify minfee calculation with reference scripts

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -516,6 +526,8 @@ data ConwayPParams f era = ConwayPParams
   , cppDRepActivity :: !(THKD ('PPGroups 'GovGroup 'NoStakePoolGroup) f EpochInterval)
   -- ^ The number of Epochs that a DRep can perform no activity without losing their @Active@ status.
+  , cppMinFeeRefScriptCoinsPerByte :: !(THKD ('PPGroups 'EconomicGroup 'SecurityGroup) f CoinPerByte)
+  -- ^ Reference scripts fee for the minimum fee calculation
```
You need to add this field in the translation to plutus here https://github.com/IntersectMBO/cardano-ledger/blob/996231f0bc288aa5fdf7cdb501b03ca782e79aea/eras/conway/impl/src/Cardano/Ledger/Conway/Plutus/Context.hs#L197

teodanciu replied:
Didn't know about this, thanks

lehins replied:
Yeah, it is a very recent addition

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/SafeHash.hs:
```diff
@@ -142,6 +143,9 @@ class SafeToHash t where
   originalBytes :: t -> ByteString
 
+  originalBytesSize :: t -> Int
+  originalBytesSize = BS.length . originalBytes
```
Overwrite this default implementation in the instance for `ShortByteString` few lines below.
This will avoid a memcopy whenever we just want to get the size

Same needs to be done with `SafeToHash` instance for `MemoBytes`

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Tools.hs:
```diff
@@ -47,14 +48,32 @@ import Lens.Micro
+  (Tx era -> Coin) ->
+  Tx era ->
+  Tx era
+_setMinFeeTx f tx =
```
Never use `_` prefix, unless you have a binding that you really need to keep around for some reason, despite that it being unused.
Underscore prefix is the instruction for the compiler to not warn you about something not being used, so it is dangerous to prefix with underscore things that are being used, because a simple spelling mistake can hide a bug. Small example that depicts a problem:
```haskell
foo _a _b _c _d _o _p = bar _a _b _c _d _c _p
```
Compiler will not warn you about the problem above

It is better to add a suffix `Internal`, if there is a need to emphasize that a function should not be for public consumption.

lehins commented on eras/babbage/impl/src/Cardano/Ledger/Babbage/UTxO.hs:
```diff
@@ -137,9 +141,17 @@ getReferenceScripts ::
+  Set (TxIn (EraCrypto era)) ->
+  [(ScriptHash (EraCrypto era), Script era)]
+getReferenceScriptsNonDistinct (UTxO mp) inputs =
+  Foldable.foldl' accum [] (Map.elems (eval (inputs ◁ mp)))
```
This will avoid the reversing artifact of foldl. Also it is a bit simpler, IMHO:
```suggestion
  [(hashScript script, script)
   | txOut <- Map.elems (eval (inputs ◁ mp))
   , SJust script <- [txOut ^. referenceScriptTxOutL]
   ]
```

teodanciu replied:
Yes, much nicer, thanks

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Wallet.hs:
```diff
@@ -446,7 +446,7 @@ getRewardProvenance globals newepochstate =
-evaluateMinFee :: EraTx era => PParams era -> Tx era -> Coin
+evaluateMinFee :: EraTx era => PParams era -> Tx era -> Int -> Coin
 evaluateMinFee = getMinFeeTx
 {-# DEPRECATED evaluateMinFee "In favor of 'getMinFeeTx'" #-}
```
We should remove this function.

Whenever type signature for a deprecated has to change, there is no longer a point to keep it around. Because the only reason deprecated functions are useful is for backwards compatibility.

Alternative would to: `evaluateMinFee pp tx = getMinFeeTx pp tx 0`, but it has been deprecated for so long that there is no longer a reason for keeping it around

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Wallet.hs:
```diff
@@ -468,6 +468,8 @@ evaluateTransactionFee ::
   -- | The number of key witnesses still to be added to the transaction.
   Word ->
+  -- | The total size of reference scripts
+  Int ->
```
Ditto. This function could be removed too, however, this is an alternative implementation that we could reuse for testing somehow. So, maybe you can put somewhere in a core testlib. I believe this function is used in one of the tests already, so it would be sad if we had to remove tests because of this.

teodanciu replied:
Indeed, it's used in two tests: `Fees` in shelley-test and `AlonzoAPI` in cardano-ledger-test.

What do you think about replacing there the usage of `evaluateTransactionFee` with `estimateMinFeeTx`  and passing zero as number of byron witnesses ? (since the former doesn't set `bootAddrTxWitsL`) 
In this way we'll have more tests for the function that we're exposing, `estimateMinFeeTx` and fewer variations of this function.

lehins replied:
> What do you think about replacing there the usage of evaluateTransactionFee with estimateMinFeeTx

Yeah, let's do that. `evaluateTransactionFee` is just dumber version of `estimateMinFeeTx`

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs:
```diff
@@ -0,0 +1,140 @@
+
+    it "a scripts referenced several times counts for each reference towards the minFee calculation" $ do
+      spendingScript <- mkScript
+      extraScripts <- sequence [mkScript, mkScript]
```
```suggestion
      extraScripts <- replicateM 2 mkScript
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs:
```diff
@@ -0,0 +1,140 @@
+
+    it "reference scripts not required for spending the input count towards the minFee calculation" $ do
+      spendingScript <- mkScript
+      extraScripts <- sequence [mkScript, mkScript, mkScript]
```
```suggestion
      extraScripts <- replicateM 3 mkScript
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs:
```diff
@@ -0,0 +1,140 @@
+          mkBasicTxBody
+            & outputsTxBodyL @era
+              .~ SSeq.fromList outs
+      let refIns = (\i -> txInAt (fromIntegral i) tx) <$> [1 .. length scripts]
```
TxIns are zero based indexed, so I think we should change how we fixup the transaction. First of all this code confused me for a moment, secondly theoretically it is possible for txOut with change to not be added.
I believe the change txOut should go at the end of the outputs, rather than a te the beginning. This will make tests more reliable IMHO.

This would cause this code to be changed to proper 0-base index:
(note that it could also be `[0..]`, since you are using it with `zip`)
```suggestion
      let refIns = (\i -> txInAt (fromIntegral i) tx) <$> [0 .. length scripts - 1]
```

teodanciu replied:
Added a commit that changes the fixup. I agree, it's more intuitive  and likely more robust.

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs:
```diff
@@ -0,0 +1,140 @@
+      -- create outputs with reference scripts and the return them mapped to their corresponding inputs
+      refScriptInToScripts <- createRefScriptsUtxos refScripts
+      -- spend the initial input by passing the reference scripts
+      tx <- spendScriptUsingRefScripts scriptSpendIn $ Map.keys refScriptInToScripts
```
You actually need a Set over in `spendScriptUsingRefScripts` 
```suggestion
      tx <- spendScriptUsingRefScripts scriptSpendIn $ Map.keysSet refScriptInToScripts
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Tx.hs:
```diff
@@ -66,6 +72,20 @@ instance Crypto c => EraTx (ConwayEra c) where
+  Int ->
+  Coin
+getConwayMinFeeTx pp tx refScriptsSize =
+  alonzoMinFeeTx pp tx <+> refScriptsSize <×> refScriptFee
```
That is one of the very few times when I will suggest redundant parenthesis :grin: 
```suggestion
  alonzoMinFeeTx pp tx <+> (refScriptsSize <×> refScriptFee)
```

teodanciu replied:
Good point, and not even hlint complains about it,

lehins replied:
> not even hlint complains about it

I think that is because those are non-standard operators that are defined in another package (instead of this module) and hlint is not aware of their infix settings. So it can't know if those parenthesis are redundant or not. 

Which is sort of the same reason why I suggested those parenthesis. It was also because those operators are non-standard and the reader might not know the precedence.

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs:
```diff
@@ -0,0 +1,140 @@
+    txInAt :: Integer -> Tx era -> TxIn (EraCrypto era)
+    txInAt index tx =
+      let txid = txIdTx tx
+       in mkTxInPartial txid index
```
It could be a useful function for testing in general, so maybe we should put it into core testlib. Also type signature could be more general and with `HasCallStack`
```suggestion
    txInAt :: (HasCallStack, Integral i) => i -> Tx era -> TxIn (EraCrypto era)
    txInAt index tx =
      let txid = txIdTx tx
       in mkTxInPartial txid (toInteger index)
```

teodanciu replied:
I have extracted it in `Test.Cardano.Ledger.Core.Utils`, but that place seems like the "miscellaneous drawer" everyone has in the house that is rarely useful. 
Couldn't find a better place in testlib. I was tempted to add it in `Core` just below the definitions of `txIdTx` and `txIdTxBody`,  to make it easier to find, but because it's using `mkTxInPartial` which has a warning to be used only in tests, I  decided against it. 
Happy to move it to better place than `Utils`

lehins replied:
Sounds like as good of a place as any, at least for the time being. Maybe we'll find a better home for it in the future.
Thank you.

lehins reviewed (approved):
This is great. Few suggestions, but nothing critical.

teodanciu committed:
Add `refScriptsSize` parameter to `Tx.getMinFeeTx`, `setMinFeeTx`

and `estimateMinFeeTx`, and use it in the calculation of Conway
`getMinFeeTx`

teodanciu committed:
Remove deprecated Wallet `evaluateMinFee`,`evaluateTransactionFee`

and use `estimateMinFeeTx` in tests where necessary.

teodanciu committed:
Add `getMinfeeTxUtxo` function to `EraUTxO` class and use it in rules

This will add the cost for the reference scripts to the computation,
starting with Conway.

teodanciu committed:
Update Imp fixup to create change output at the end at the list

in order to make tests less fragile.
For this, instead of storing only `rootTxId` in `ImpTestState`, we need
to store the whole TxIn as `rootTxIn`.

teodanciu committed:
Extract `txInAt` utility function in core testlib `Utils`

and use it to avoid duplication

lehins commented on eras/conway/impl/CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ## 1.13.0.0
 
+* Add `cppMinFeeRefScriptCoinsPerByte` to `ConwayPParams` and `cppMinFeeRefScriptCoinsPerByteL`
```
```suggestion
* Add `cppMinFeeRefScriptCoinsPerByte` to `ConwayPParams` and `ppMinFeeRefScriptCoinsPerByteL`
```
